File: ocean_reward_delta.py
#!/usr/bin/env python3
import requests
import sys

def get_ocean_health():
    slug = "ocean"
    pool_url = f"https://mempool.space/api/v1/mining/pool/{slug}"
    blocks_url = f"https://mempool.space/api/v1/mining/pool/{slug}/blocks"
    price_url = "https://mempool.space/api/v1/prices"

    try:
        # 1. Fetch Current BTC Price (USD)
        price_res = requests.get(price_url, timeout=10)
        price_res.raise_for_status()
        btc_usd = price_res.json().get("USD")

        # 2. Get Aggregate Pool Health
        pool_res = requests.get(pool_url, timeout=10)
        pool_res.raise_for_status()
        pool_data = pool_res.json()
        avg_health = pool_data.get("avgBlockHealth", "N/A")

        # 3. Get Individual Block Data
        block_res = requests.get(blocks_url, timeout=10)
        block_res.raise_for_status()
        blocks = block_res.json()

        print(f"--- OCEAN Pool Metrics (BTC Price: ${btc_usd:,.2f}) ---")
        print(f"Aggregate Pool Health: {avg_health}%")
        print("\n" + "-" * 90)
        print(f"{'Height':<10} | {'Health':<8} | {'Actual (Sats)':<15} | {'Expected (Sats)':<15} | {'Loss (USD)':<10}")
        print("-" * 90)

        for b in blocks[:10]: # Limiting display to 10 for readability
            height = b.get("height")
            extras = b.get("extras", {})

            match_rate = extras.get("matchRate", 0)
            actual_reward = extras.get("reward", 0)
            # Calculate Expected: Actual Reward / (Match Rate / 100)
            # This accounts for the fees missed due to template mismatch
            if match_rate > 0:
                expected_reward = int(actual_reward / (match_rate / 100))
            else:
                expected_reward = actual_reward

            diff_sats = expected_reward - actual_reward
            diff_usd = (diff_sats / 100_000_000) * btc_usd

            print(f"{height:<10} | {match_rate:>6.2f}% | {actual_reward:>15,} | {expected_reward:>15,} | ${diff_usd:>8.2f}")

    except Exception as e:
        print(f"Error fetching metrics: {e}")
        sys.exit(1)

if __name__ == "__main__":
    # Use the interpreter from the venv if available
    get_ocean_health()

File: src/bin/ocean_reward_delta.rs
use anyhow::Result;
use ocean_loss_estimator_rs::models::{Block, BlockExtras, PoolData};
use ocean_loss_estimator_rs::utils::fetch_from_mirror;
use serde::Deserialize; // Still needed for CurrentPrice
use serde_json::Value;

#[derive(Debug, Deserialize)]
struct CurrentPrice {
    #[serde(rename = "USD")]
    usd: f64,
}

async fn get_ocean_reward_delta_rust() -> Result<()> {
    let slug = "ocean";
    let pool_path = format!("/api/v1/mining/pool/{}", slug);
    let blocks_path = format!("/api/v1/mining/pool/{}/blocks", slug);
    let price_path = "/api/v1/prices";

    // 1. Fetch Current BTC Price (USD)
    let price_data: Value = fetch_from_mirror(&price_path, 0, 10).await?;
    let price_res: CurrentPrice = serde_json::from_value(price_data)?;
    let btc_usd = price_res.usd;

    // 2. Get Aggregate Pool Health
    let pool_data: Value = fetch_from_mirror(&pool_path, 0, 10).await?;
    let pool_res: PoolData = serde_json::from_value(pool_data)?;
    let avg_health = pool_res.avg_block_health.unwrap_or_default();

    // 3. Get Individual Block Data
    let blocks_data: Value = fetch_from_mirror(&blocks_path, 0, 10).await?;
    let block_res: Vec<Block> = serde_json::from_value(blocks_data)?;

    println!("--- OCEAN Pool Metrics (BTC Price: ${:.2}) ---", btc_usd);
    println!("Aggregate Pool Health: {}%", avg_health);
    println!("
{:->90}", "");
    println!("{:<10} | {:<8} | {:<15} | {:<15} | {:<10}",
             "Height", "Health", "Actual (Sats)", "Expected (Sats)", "Loss (USD)");
    println!("{:->90}", "");

    for b in block_res.iter().take(10) { // Limiting display to 10 for readability
        let height = b.height;
        let extras = b.extras.as_ref().unwrap_or(&BlockExtras { match_rate: Some(0.0), reward: Some(0), expected_fees: Some(0) });

        let match_rate = extras.match_rate.unwrap_or(0.0);
        let actual_reward = extras.reward.unwrap_or(0);

        let expected_reward = if match_rate > 0.0 {
            (actual_reward as f64 / (match_rate / 100.0)) as u64
        } else {
            actual_reward
        };

        let diff_sats = expected_reward.saturating_sub(actual_reward);
        let diff_usd = (diff_sats as f64 / 100_000_000.0) * btc_usd;

        println!("{:<10} | {:>6.2}% | {:>15} | {:>15} | ${:>8.2}",
                 height, match_rate, actual_reward, expected_reward, diff_usd);
    }

    Ok(())
}

#[tokio::main]
async fn main() -> Result<()> {
    get_ocean_reward_delta_rust().await
}
